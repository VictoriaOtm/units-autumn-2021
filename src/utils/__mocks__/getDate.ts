export const getDate = (date: number): string => {
	return new Date(date).toISOString();
};
