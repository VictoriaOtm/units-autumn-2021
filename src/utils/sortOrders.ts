// Array sort doc
// @see https://developer.mozilla.org/ru/docs/Web/JavaScript/Reference/Global_Objects/Array/sort#Description

import {Order} from '../data/fakeOrders';

export const sortTypes = {
	DATE: 'date',
	COUNT: 'count',
};

/**
 * Функция сортировки заказов
 * @param {Array} orders - массив заказов
 * @param {Function} sortFunction - функция сортировки
 */
export const sortOrders = (orders: Order[], sortFunction: any) => {
	if (!orders || !orders.length) {
		return;
	}

	if (!sortFunction || typeof sortFunction !== 'function') {
		return;
	}

	orders.sort(sortFunction);
};

/**
 * Функция возвращает нужный колбэк
 * по переданному типу сортировки
 * @param sortType
 * @returns {sortByItemCount|sortByDate}
 */
export const getSortFunction = (sortType: typeof sortTypes.COUNT | typeof sortTypes.DATE) => {
	switch(sortType) {
	case sortTypes.DATE:
		return sortByDate;
	case sortTypes.COUNT:
		return sortByItemCount;
	}
};


/**
 * Колбэк для сортировки массива заказов
 * Сравнивает 2 заказа по количеству товаров в заказе
 * по возрастанию
 * @param order1
 * @param order2
 */
export const sortByItemCount = (order1: Order, order2: Order): number => {
	const {items: items1} = order1;
	const {items: items2} = order2;

	if (!items1 || !items2) {
		return 0;
	}

	switch (true) {
	case items1.length < items2.length:
		return -1;
	case items1.length === items2.length:
		return 0;
	default:
		return 1;
	}
};

/**
 * Колбэк для сортировки массива заказов
 * Сравнивает 2 заказа по дате
 * по убыванию (сначала новые заказы)
 * @param order1
 * @param order2
 */
export const sortByDate = (order1: Order, order2: Order): number => {
	const {date: date1} = order1;
	const {date: date2} = order2;

	if (!date1 || !date2) {
		return 0;
	}

	switch (true) {
	case date1 < date2:
		return 1;
	case date1 > date2:
		return -1;
	default:
		return 0;
	}
};
