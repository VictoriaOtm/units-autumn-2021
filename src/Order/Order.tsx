import React from 'react';
import './Order.css';
import {getDate} from '../utils/getDate';
import {Order} from '../data/fakeOrders';

export type OrderComponentProps = {
	order: Order|null,
}

export const OrderComponent: React.FC<OrderComponentProps> = ({
	order
}) => {
	if (!order || !order.shop || !order.date) {
		return null;
	}

	const getItems = () => {
		const {items} = order;

		if (!items || !items.length) {
			return (
				<div>
					{'Заказ пуст...'}
				</div>
			);
		}

		return items.map((item, index) => (
			<div
				key={index}
				className='Order-item'
			>
				{item}
			</div>
		));
	};

	return (
		<div className='Order'>
			<div className='Order-header'>
				<span className='Order-shop'>{order.shop}</span>
				<span id='date'>{getDate(order.date)}</span>
			</div>

			<div className='Order-items'>
				{getItems()}
			</div>
		</div>
	);
};
