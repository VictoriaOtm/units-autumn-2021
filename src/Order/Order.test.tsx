import React from 'react';
import { OrderComponent } from './Order';
import { Order } from '../data/fakeOrders';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

jest.mock('../utils/getDate')
import { getDate } from '../utils/getDate';

describe('Order.tsx', () => {
	afterAll(() => {
		jest.resetAllMocks();
	});

	it('empty', () => {
		const testOrder: Order = {};
		var wrapper = shallow(<OrderComponent order={testOrder} />);
		expect(wrapper).toMatchSnapshot()
	});

	it('no items', () => {
		const testOrder: Order = {
			date: 1,
			shop: 'shop'
		};
		var wrapper = shallow(<OrderComponent order={testOrder} />);
		expect(wrapper).toMatchSnapshot()
	});

	it('usual order', () => {
		getDate.mockReturnValue('Today');
		const testOrder: Order = {
			items: ['abc'],
			date: 1,
			shop: 'shop'
		};


		var wrapper = shallow(<OrderComponent order={testOrder} />);
		expect(wrapper).toMatchSnapshot()
		expect(getDate).toHaveBeenCalled()
		jest.clearAllMocks()
	});
});

