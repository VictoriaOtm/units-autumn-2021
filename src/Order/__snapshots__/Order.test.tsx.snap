// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Order.tsx empty 1`] = `""`;

exports[`Order.tsx no items 1`] = `
<div
  className="Order"
>
  <div
    className="Order-header"
  >
    <span
      className="Order-shop"
    >
      shop
    </span>
    <span
      id="date"
    />
  </div>
  <div
    className="Order-items"
  >
    <div>
      Заказ пуст...
    </div>
  </div>
</div>
`;

exports[`Order.tsx usual order 1`] = `
<div
  className="Order"
>
  <div
    className="Order-header"
  >
    <span
      className="Order-shop"
    >
      shop
    </span>
    <span
      id="date"
    >
      Today
    </span>
  </div>
  <div
    className="Order-items"
  >
    <div
      className="Order-item"
      key="0"
    >
      abc
    </div>
  </div>
</div>
`;
